shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Deleg.hs

```diff
@@ -220,43 +220,48 @@ delegationTransition = do
   case c of
     DCertDeleg (RegKey hk) -> do
       -- note that pattern match is used instead of regCred, as in the spec
-      hk ∉ dom (_stkCreds ds) ?! StakeKeyAlreadyRegisteredDELEG hk
+      not (haskey hk (_stkCreds ds)) ?! StakeKeyAlreadyRegisteredDELEG hk
 
       pure $
         ds
-          { _stkCreds = _stkCreds ds ∪ singleton hk slot,
-            _rewards = _rewards ds ∪ Map.singleton (RewardAcnt network hk) (Coin 0), -- ∪ is override left
-            _ptrs = _ptrs ds ∪ Map.singleton ptr hk
+          { _stkCreds = addpair hk slot (_stkCreds ds), -- _stkCreds ds ∪ (singleton hk slot)
+            _rewards = addpair (RewardAcnt network hk) (Coin 0) (_rewards ds), -- _rewards ds ∪ (singleton (RewardAcnt network hk) (Coin 0) )
+            _ptrs = addpair ptr hk (_ptrs ds) -- _ptrs ds ∪ (singleton ptr hk)
           }
     DCertDeleg (DeRegKey hk) -> do
       -- note that pattern match is used instead of cwitness, as in the spec
-      hk ∈ dom (_stkCreds ds) ?! StakeKeyNotRegisteredDELEG hk
+      haskey hk (_stkCreds ds) ?! StakeKeyNotRegisteredDELEG hk
 
       let rewardCoin = Map.lookup (RewardAcnt network hk) (_rewards ds)
       rewardCoin == Just 0 ?! StakeKeyNonZeroAccountBalanceDELEG rewardCoin
 
       pure $
         ds
-          { _stkCreds = Set.singleton hk ⋪ _stkCreds ds,
-            _rewards = Set.singleton (RewardAcnt network hk) ⋪ _rewards ds,
-            _delegations = Set.singleton hk ⋪ _delegations ds,
+          { _stkCreds = removekey hk (_stkCreds ds),
+            _rewards = removekey (RewardAcnt network hk) (_rewards ds),
+            _delegations = removekey hk (_delegations ds),
             _ptrs = _ptrs ds ⋫ Set.singleton hk
+            -- TODO make _ptrs a bijection. This operation takes time proportional to (_ptrs ds)
+            -- OR turn _stkCreds into a mapping of stake credentials to pointers
+            -- note that the slot values in _stkCreds is no longer needed (no decay)
+            -- then we could use (lookup hk (_delecations ds)) to get ptr, then use domain removal on (_ptrs ds) rather than range removal
           }
     DCertDeleg (Delegate (Delegation hk dpool)) -> do
       -- note that pattern match is used instead of cwitness and dpool, as in the spec
-      hk ∈ dom (_stkCreds ds) ?! StakeDelegationImpossibleDELEG hk
+      haskey hk (_stkCreds ds) ?! StakeDelegationImpossibleDELEG hk
```

Comment:
I would definitely be interested in investigating the use of rewrite rules here... the original syntax is much closer to the spec

Reply:
I have added some comments to indicate this.